### PR TITLE
Fix a memory leak in the resize images command

### PR DESCRIPTION
**Pull request by ausi**
This implements the fix for #849 as described in https://github.com/contao/contao/issues/849#issuecomment-546679544

Todo:

- [x] Rework the console output
- [x] Implement `--throttle` and `--time-limit`
- [x] Update tests

See #877

### CODE REVIEW

**Comment by leofeyer** [comment]
@ausi Are you still working on this one?

**Comment by ausi** [comment]
Yes.

File: core-bundle/src/Command/ResizeImagesCommand.php [hunk 1]
```diff
@@ -85,181 +102,250 @@ protected function configure(): void
     {
         $this
             ->addOption('time-limit', 'l', InputOption::VALUE_OPTIONAL, 'Time limit in seconds', '0')
-            ->addOption('concurrent', 'c', InputOption::VALUE_OPTIONAL, 'Run multiple processes concurrently', '1')
-            ->addOption('throttle', 't', InputOption::VALUE_OPTIONAL, 'Pause between resizes to limit CPU utilization, 0.1 relates to 10% CPU usage', '1')
+            ->addOption('concurrent', 'c', InputOption::VALUE_OPTIONAL, 'Run multiple processes concurrently with a value larger than 1 or pause between resizes to limit CPU utilization with values lower than 1.0', '1')
+            ->addOption('throttle', 't', InputOption::VALUE_OPTIONAL, '(Deprecated) Use the concurrent option instead', '1')
```
Comment (ausi) [inline core-bundle/src/Command/ResizeImagesCommand.php]: Is there a way, or do we have a convention on how to deprecate command options?

**Comment by ausi** [comment]
Besideds missing unit tests, this is ready for review 🎉

File: core-bundle/src/Command/ResizeImagesCommand.php (hunk 1 above)
Comment (Toflar) [reply core-bundle/src/Command/ResizeImagesCommand.php]: Nope. Would be a great addition to the console component :)

File: core-bundle/src/Command/ResizeImagesCommand.php [hunk 2]
```diff
@@ -85,181 +102,250 @@ protected function configure(): void
...
-        foreach ($this->storage->listPaths() as $path) {
-            $sleep = (1 - $throttle) / $throttle * $lastDuration;
+        $timeLimit = (float) $input->getOption('time-limit');
+        $throttle = (float) $input->getOption('throttle');
+        $concurrent = (float) $input->getOption('concurrent');
 
-            if ($timeLimit && microtime(true) - $startTime + $maxDuration + $sleep > $timeLimit) {
-                $io->writeln("\n".'Time limit of '.$timeLimit.' seconds reached.');
+        if (1.0 !== $throttle) {
+            @trigger_error('Using the throttle option is deprecated and will no longer work in Contao 5.0. Use the concurrent option instead.', E_USER_DEPRECATED);
```
Comment (Toflar) [inline core-bundle/src/Command/ResizeImagesCommand.php:133]: This would need to be adjusted when merging upstream to 4.10 or master @leofeyer. I'm not sure when the switch to `trigger_deprecation()` came.

**Review by Toflar — COMMENTED** [review]
Looks good to me! Hopefully I can soon include that into my cronjobs again 🎉

File: core-bundle/src/Command/ResizeImagesCommand.php (hunk 2 above)
Comment (leofeyer) [reply core-bundle/src/Command/ResizeImagesCommand.php:133]: Yes, 4.10.

File: core-bundle/src/Command/ResizeImagesCommand.php [hunk 3]
```diff
@@ -85,181 +102,250 @@ protected function configure(): void
...
-            throw new InvalidArgumentException(sprintf('Time-limit value "%s" is invalid.', $timeLimit));
-        }
-
-        if ($throttle < 0.001 || $throttle > 1) {
-            throw new InvalidArgumentException(sprintf('Throttle value "%s" is invalid.', $throttle));
-        }
-
-        if ($concurrent < 1) {
-            throw new InvalidArgumentException(sprintf('Concurrent value "%s" is invalid.', $concurrent));
+        if (!$output instanceof ConsoleOutputInterface) {
```
Comment (m-vo) [inline core-bundle/src/Command/ResizeImagesCommand.php:119]: Why is this needed? Imo none of the `ConsoleOutputInterface`'s methods are in use.

File: core-bundle/src/Command/ResizeImagesCommand.php [hunk 4]
```diff
@@ -85,181 +102,250 @@ protected function configure(): void
...
+        return $this->resizeImages($timeLimit, $concurrent);
     }
 
-    private function resizeImage(string $path, SymfonyStyle $io): float
+    private function resizeImage(string $path, bool $quiet = false): int
     {
-        $startTime = microtime(true);
-
-        if ($this->filesystem->exists($path)) {
+        if ($this->filesystem->exists($this->targetDir.'/'.$path)) {
```
Comment (m-vo) [inline core-bundle/src/Command/ResizeImagesCommand.php:165]: This line + below should probably also be converted into a `Path::join()` + variable when merging upstream. :slightly_smiling_face: 

/cc @leofeyer

File: core-bundle/src/Command/ResizeImagesCommand.php [hunk 5]
```diff
@@ -85,181 +102,250 @@ protected function configure(): void
...
-            $processes
-        ));
+        foreach ($this->storage->listPaths() as $path) {
+            while (true) {
+                if ($timeLimit && microtime(true) - $startTime > $timeLimit) {
+                    $this->io->writeln('Time limit of '.$timeLimit.' seconds reached.');
+                    break 2;
+                }
+
+                foreach ($processes as $index => $process) {
```
Comment (m-vo) [inline core-bundle/src/Command/ResizeImagesCommand.php:280]: Do I get this correctly? This implements a process pool of `n = concurrent` processes (if `concurrent >= 1`) and otherwise adds throttling of `m * elapsedMicroSeconds` (with `m >= 1` for `concurrent >= 0.5`, `0 < m < 1` else) between starting a new process. 

- Why does the wait time scale with the elapsed time? There has only be one parallel process that was waited on before continuing. Or am I missing something?
- Why do we need the additional `usleep(15000)`?
- The `$startTimes` array can be omitted I guess. The single usage in L285 could imo be `$startTime` instead.

File: core-bundle/src/Command/ResizeImagesCommand.php (hunk 3 above)
Comment (ausi) [reply core-bundle/src/Command/ResizeImagesCommand.php:119]: `$output->section()` is in use. But I just now noticed that the usage is now above this check 🙈

File: core-bundle/src/Command/ResizeImagesCommand.php (hunk 5 above)
Comment (ausi) [reply core-bundle/src/Command/ResizeImagesCommand.php:280]: > Why does the wait time scale with the elapsed time?

Because we want to distribute the CPU utilization. With `0.5` we should wait the same time as the last resize took to get an average of 50% CPU time.

> Why do we need the additional `usleep(15000)`?

So that the PHP process doesn’t go crazy while waiting on the sub processes.

> The `$startTimes` array can be omitted I guess. The single usage in L285 could imo be `$startTime` instead.

Yes, I agree.

Comment (ausi) [reply core-bundle/src/Command/ResizeImagesCommand.php:280]: Fixed in 7342a2822a250b0abca29b9525fc0209fa4eaa10

File: core-bundle/src/Command/ResizeImagesCommand.php (hunk 3 above)
Comment (ausi) [reply core-bundle/src/Command/ResizeImagesCommand.php:119]: Fixed in 54c1e902208d58505e1eac0eeb4b8d31f718b1ff

File: core-bundle/src/Command/ResizeImagesCommand.php [hunk 6]
```diff
@@ -85,8 +102,9 @@ protected function configure(): void
     {
         $this
             ->addOption('time-limit', 'l', InputOption::VALUE_OPTIONAL, 'Time limit in seconds', '0')
-            ->addOption('concurrent', 'c', InputOption::VALUE_OPTIONAL, 'Run multiple processes concurrently', '1')
-            ->addOption('throttle', 't', InputOption::VALUE_OPTIONAL, 'Pause between resizes to limit CPU utilization, 0.1 relates to 10% CPU usage', '1')
+            ->addOption('concurrent', 'c', InputOption::VALUE_OPTIONAL, 'Run multiple processes concurrently with a value larger than 1 or pause between resizes to limit CPU utilization with values lower than 1.0', '1')
```
Comment (m-vo) [inline core-bundle/src/Command/ResizeImagesCommand.php:105]: Would it make sense to increase the default value? Or is a higher parallelization already a problem on most hostings?

Comment (ausi) [reply core-bundle/src/Command/ResizeImagesCommand.php:105]: I think so, yes. On most hosting packages you should probably use a value less than `0.5`.

**Review by Toflar — DISMISSED** [review]

File: core-bundle/src/Command/ResizeImagesCommand.php [hunk 7]
```diff
@@ -97,169 +116,240 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
+        }
+
         $timeLimit = (float) $input->getOption('time-limit');
         $throttle = (float) $input->getOption('throttle');
-        $concurrent = (int) $input->getOption('concurrent');
+        $concurrent = (float) $input->getOption('concurrent');
 
-        if ($timeLimit < 0) {
-            throw new InvalidArgumentException(sprintf('Time-limit value "%s" is invalid.', $timeLimit));
+        if (1.0 !== $throttle) {
```
Comment (leofeyer) [inline core-bundle/src/Command/ResizeImagesCommand.php]: Should't we check `$input->hasOption()` instead?

**Comment by m-vo** [comment]
I'll get a lot of "Unable to open image [...]" errors when running it with `--no-sub-process`. 

The actual exception is thrown when `Imagick/Imagine` calls `new \Imagick($path)` and says "cache resources exhausted".

> ImagickException, code 445
> /[..]/vendor/imagine/imagine/src/Imagick/Imagine.php L69
>
> cache resources exhausted `[...]/files/public/images/QbjkOU9UZb0.jpg' @ error/cache.c/OpenPixelCache/3984

I run this 5 times in a row now with identical outcome: 200 images in a gallery, 600 resizes are executed. They run fast until ~335, then way slower till the end. Always at image 556 the first exception is thrown.

Not sure if this is a system config issue or if the "cache" can be cleared at some time.

**Comment by ausi** [comment]
> They run fast until ~335

Thats probably when you hit the memory limit and IMagick starts to use the disk instead.

> says "cache resources exhausted"

Here the disk limit seems to be reached as well.

I did not find a way to fix the memory leak or how to programmatically free the resources of IMagick.

Without `--no-sub-process` everything works fine and all images get resized? This would prove that this PR works ☺️🎉

File: core-bundle/src/Command/ResizeImagesCommand.php (hunk 7 above)
Comment (ausi) [reply core-bundle/src/Command/ResizeImagesCommand.php]: I don’t think so because `$input->hasOption()` would always return true because the option exists in the definition of the command.

**Comment by Toflar** [comment]
Imho that's the whole point of this PR, the memory should be freed once the subprocess is killed/finished, no?

File: core-bundle/src/Command/ResizeImagesCommand.php (hunk 7 above)
Comment (leofeyer) [reply core-bundle/src/Command/ResizeImagesCommand.php]: Ah, you are right. Then we should probably check `null === $input->getOption()`, because if someone uses the option with `1`, the deprecation warning will not be triggered. 🙈

Comment (ausi) [reply core-bundle/src/Command/ResizeImagesCommand.php]: Done in 54d17c6a6726fe159c78c5907a085378561acd82

**Comment by Toflar** [comment]
I wonder if we should even allow this option or just abort and display an error if subprocesses aren't supported? There's no point in providing an option that may work depending on your memory and the number of files...sounds like russian roulette to me.

**Comment by ausi** [comment]
If somone uses GD, `--no-sub-process` can save a great amount of time, also I need it for the unit tests ☺️

**Comment by Toflar** [comment]
I see. I still think there should be some kind of warning being displayed so that if you run into such messages, you know at least why this is and what you can do (make sure subprocesses can be started, whatever that means :)). I like self-documenting commands :)

**Comment by ausi** [comment]
Done in 9bfbaaab2f4bbdb1664beb27b129cdee0715c348

**Review by Toflar — APPROVED** [review]

**Review by m-vo — APPROVED** [review]
Really cool, I like it! :tada:

**Comment by leofeyer** [comment]
Thank you @ausi.

File: core-bundle/src/Command/ResizeImagesCommand.php (hunk 4 above)
Comment (leofeyer) [reply core-bundle/src/Command/ResizeImagesCommand.php:165]: See #2041.

File: core-bundle/src/Command/ResizeImagesCommand.php (hunk 2 above)
Comment (leofeyer) [reply core-bundle/src/Command/ResizeImagesCommand.php:133]: See #2041.
